google/cloud/internal/oauth2_impersonate_service_account_credentials.cc

```diff
@@ -34,6 +37,88 @@ GenerateAccessTokenRequest MakeRequest(
 
 }  // namespace
 
+StatusOr<ImpersonatedServiceAccountCredentialsInfo>
+ParseImpersonatedServiceAccountCredentials(std::string const& content,
+                                           std::string const& source) {
+  auto credentials = nlohmann::json::parse(content, nullptr, false);
+  if (credentials.is_discarded()) {
+    return internal::InvalidArgumentError(
+        "Invalid ImpersonateServiceAccountCredentials, parsing failed on data "
+        "from " +
+            source,
+        GCP_ERROR_INFO());
+  }
+
+  ImpersonatedServiceAccountCredentialsInfo info;
+
+  auto it = credentials.find("service_account_impersonation_url");
+  if (it == credentials.end()) {
+    return internal::InvalidArgumentError(
+        "Missing `service_account_impersonation_url` field on data from " +
+            source,
+        GCP_ERROR_INFO());
+  }
+  if (!it->is_string()) {
+    return internal::InvalidArgumentError(
+        "Malformed `service_account_impersonation_url` field is not a string "
+        "on data from " +
+            source,
+        GCP_ERROR_INFO());
+  }
+  // We strip the service account from the path URL.
+  auto url = it->get<std::string>();
+  auto service_account_resource =
+      absl::StripSuffix(url, ":generateAccessToken");
```

Comment:
We will also want to support `generateIdToken` for `universe_domain`.

Reply:
The only difference between `generateIdToken` and `generateAccessToken` is the action name in the url.

Reply:
Thanks for bringing this up. I didn't know there were other cases.

> The only difference ...

Hmm are you sure? I imagine this would have to change:

https://github.com/googleapis/google-cloud-cpp/blob/58a05d0c89d505131e2ccb99b03b29575c5c25a5/google/cloud/internal/oauth2_impersonate_service_account_credentials.cc#L50-L53

https://github.com/googleapis/google-cloud-cpp/blob/58a05d0c89d505131e2ccb99b03b29575c5c25a5/google/cloud/internal/oauth2_minimal_iam_credentials_rest.cc#L55

https://github.com/googleapis/google-cloud-cpp/blob/58a05d0c89d505131e2ccb99b03b29575c5c25a5/google/cloud/internal/oauth2_minimal_iam_credentials_rest.cc#L74-L80

I am tempted to defer support for `generateIdToken` to a later PR. WDYT?

Reply:
What I am thinking is, we only need to support `generateIdToken` for the new ADC credential format(type `impersonated_service_account`), because it specified `service_account_impersonation_url` which tells us it expects using `generateIdToken`.

For the ADC credentials with type `external_account` or `service_account` , we should keep using `generateAccessToken`, because it doesn't specify `generateIdToken` anywhere.

We could directly pass `service_account_impersonation_url` as the endpoint in oauth2_minimal_iam_credentials_rest.cc instead of making it by `universe_domain` and `service_account`.

Reply:
It is not as simple as swapping out the path. They are different RPCs with different request types. We would also have to account for the different request.

https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateIdToken

vs.

https://cloud.google.com/iam/docs/reference/credentials/rest/v1/projects.serviceAccounts/generateAccessToken

---

What I am unsure about is if the returned ID token plugs in the same way to the `Authorization: Bearer <token>` header. Do you know?

Reply:
Yes, I will ping you the doc link about it. I could trust the doc and test with the credential it provides us.

Reply:
Added a comment that says why we only handle `generateAccessToken` for now, and set us up a little better to handle `generateIdToken` in the future.

PTAL